connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java

```diff
@@ -357,21 +358,53 @@ public URI advertisedUrl() {
         ServerConnector serverConnector = findConnector(advertisedSecurityProtocol);
         builder.scheme(advertisedSecurityProtocol);
 
-        String advertisedHostname = config.getString(WorkerConfig.REST_ADVERTISED_HOST_NAME_CONFIG);
-        if (advertisedHostname != null && !advertisedHostname.isEmpty())
-            builder.host(advertisedHostname);
-        else if (serverConnector != null && serverConnector.getHost() != null && serverConnector.getHost().length() > 0)
-            builder.host(serverConnector.getHost());
+        String hostNameOverride = hostNameOverride(serverConnector);
+        if (hostNameOverride != null) {
+            builder.host(hostNameOverride);
+        }
 
         Integer advertisedPort = config.getInt(WorkerConfig.REST_ADVERTISED_PORT_CONFIG);
         if (advertisedPort != null)
             builder.port(advertisedPort);
         else if (serverConnector != null && serverConnector.getPort() > 0)
             builder.port(serverConnector.getPort());
 
-        log.info("Advertised URI: {}", builder.build());
+        URI uri = builder.build();
+        maybeThrowInvalidHostNameException(uri, hostNameOverride);
+        log.info("Advertised URI: {}", uri);
 
-        return builder.build();
+        return uri;
+    }
+
+    private String hostNameOverride(ServerConnector serverConnector) {
+        String advertisedHostname = config.getString(WorkerConfig.REST_ADVERTISED_HOST_NAME_CONFIG);
+        if (advertisedHostname != null && !advertisedHostname.isEmpty())
+            return advertisedHostname;
+        else if (serverConnector != null && serverConnector.getHost() != null && serverConnector.getHost().length() > 0)
+            return serverConnector.getHost();
+        return null;
+    }
+
+    /**
+     * Parses the uri and throws a more definitive error
+     * when the internal node to node communication can't happen due to an invalid host name.
+     */
+    static void maybeThrowInvalidHostNameException(URI uri, String hostNameOverride) {
+        //java URI parsing will fail silently returning null in the host if the host name contains invalid characters like _
+        //this bubbles up later when the Herder tries to communicate on the advertised url and the current HttpClient fails with an ambiguous message
```

Comment:
It's not that URI parsing _fails_, it actually succeeds, but it doesn't result in a server-based authority (it's a registry-based authority instead, precisely because what would be the hostname part cannot be a hostname because of the illegal syntax) and so there really is no host. Can we amend the comment to something like this:

```suggestion
        // When the hostname contains illegal characters (e.g. _) the URI authority will be registry-based, lacking a host
        // so detect it here to prevent a confusing error later when the Herder tries to communicate on the advertised url
```